### DISCUSSION TITLE
weird interaction between asdf, direnv, fish, and the catalina mac app store

### DISCUSSION THREAD

**Discussion by will** [asdf-community/asdf-direnv#59] [Q&A] [answered]
While maybe someone here knows the answer, I'm mostly posting this because I can't find information on it anywhere, and maybe this will eventually be indexed by search engines and over time we'll all figure it out eventually. Or at the very least I'll find this if I run into the problem again after having forgotten everything.

I was having problems both installing updates and new software on the mac app store. The ring around the download icon would get to ~75% full but not progress further. I tried some things like kill, kill -9 the mac app store and processes that looked related, and also rebooted, but that did not fix things.

I noticed that bash was high in cpu usage. I knew this must be some system process since I use fish as my login shell. I killed that bash process, but another one would pop up.

I looked into it with htop and found a process tree that had `package_script_s ` running as root, then underneath it perl execing another perl, which was sudoing into my user with a command something like `rm -R ~/Library/Caches/com.apple.helpd` and under that a stack of several `bash /usr/local/opt/asdf/bin/asdf exec direnv hook fish` processes that seemed to be launching more up to a limit, then the last one would disappear, then reappear after a short time.

I ignored the whole asdf thing at this time, and manually ran the `rm` it was trying to, and it moved onto another `rm`, this time `~/Library/Preferences/com.apple.help.plist`. 

At this point I actually suspected the asdf direnv thing, and commented out these lines from my fish config

``` fish
if test $SHLVL = 1
  source /usr/local/opt/asdf/asdf.fish # too slow original: source (brew --prefix asdf)/asdf.fish
  set -gp fish_user_paths "$HOME/.asdf/bin"
  eval (direnv hook fish)
end
```

And killed the parent process of all of the bash things. At this point all of the App Store installs and updates worked fine and finished. 

There were still some bash asdf things around, probably from when I was trying to kill things and didn't do it property. I have a screenshot of htop at that point (but not in the original state with the parent perl processes)
<img width="1214" alt="image" src="https://user-images.githubusercontent.com/1973/94274477-0979eb00-ff46-11ea-8e76-4052cfae2daf.png">

I would imagine the proper fix would be for apple's install scripts to use sudo with flags to not launch my profile. I doubt they were written to be resilient to such a strange shell as fish, and random env manipulating things like direnv on top of that. However that is out of my control, so I wonder if maybe I can do some sort of hack to look at the shell's parent process and only do the asdf and/or direnv stuff if it's being run in a tty or as part of iterm2 or terminal. Something like that maybe.

**Comment by smorimoto** [comment, marked as answer]
I don't know because I don't have a macOS environment, but it is quite weird.

By the way, your fish configuration is not recommended. Please refer to my config.
https://github.com/smorimoto/dotfiles/blob/6d607966bc6e7d918fdbbd1d3458ab316ed429f5/config/fish/config.fish#L1-L7
